agent-info/src/fetcher.rs

```diff
@@ -0,0 +1,345 @@
+use ddcommon::{connector::Connector, Endpoint};
+use hyper::{self, body::Buf, header::HeaderName};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::sleep;
+
+#[allow(clippy::declare_interior_mutable_const)]
+const DATADOG_AGENT_STATE: HeaderName = HeaderName::from_static("datadog-agent-state");
+
+#[derive(Debug)]
+enum FetchInfoStatus {
+    SameState,
+    NewState(Box<AgentInfo>),
+}
+
+/// Fetch info from the given info_endpoint and compare its state to the current state hash.
+///
+/// If the state hash is different from the current one:
+/// - Return a `FetchInfoStatus::NewState` of the info struct
+/// - Else return `FetchInfoStatus::SameState`
+async fn fetch_info_with_state(
+    info_endpoint: &Endpoint,
+    current_state_hash: Option<&str>,
+) -> Result<FetchInfoStatus> {
+    let req = info_endpoint
+        .into_request_builder(concat!("Libdatadog/", env!("CARGO_PKG_VERSION")))?
+        .method(hyper::Method::GET)
+        .body(hyper::Body::empty());
+    let client = hyper::Client::builder().build(Connector::default());
+    let res = client.request(req?).await?;
+    let new_state_hash = res
+        .headers()
+        .get(DATADOG_AGENT_STATE)
+        .ok_or_else(|| anyhow!("Agent state header not found"))?
+        .to_str()?;
+    if current_state_hash.is_some_and(|state| state == new_state_hash) {
+        return Ok(FetchInfoStatus::SameState);
+    }
+    let state_hash = new_state_hash.to_string();
+    let body_bytes = hyper::body::aggregate(res.into_body()).await?;
+    let info = Box::new(AgentInfo {
+        state_hash,
+        info: serde_json::from_reader(body_bytes.reader())?,
+    });
+    Ok(FetchInfoStatus::NewState(info))
+}
+
+/// Fetch the info endpoint once and return the info.
+///
+/// Can be used for one-time access to the agent's info. If you need to access the info several
+/// times use `AgentInfoFetcher` to keep the info up-to-date.
+///
+/// # Example
+/// ```no_run
+/// # use anyhow::Result;
+/// # #[tokio::main]
+/// # async fn main() -> Result<()> {
+/// // Define the endpoint
+/// let endpoint = ddcommon::Endpoint::from_url("http://localhost:8126/info".parse().unwrap());
+/// // Fetch the info
+/// let agent_info = agent_info::fetch_info(&endpoint).await.unwrap();
+/// println!("Agent version is {}", agent_info.info.version.unwrap());
+/// # Ok(())
+/// # }
+/// ```
+pub async fn fetch_info(info_endpoint: &Endpoint) -> Result<Box<AgentInfo>> {
+    match fetch_info_with_state(info_endpoint, None).await? {
+        FetchInfoStatus::NewState(info) => Ok(info),
+        FetchInfoStatus::SameState => Err(anyhow!("Invalid state header")),
+    }
+}
+
+/// Fetch the info endpoint and update an ArcSwap keeping it up-to-date.
+///
+/// Once the fetcher has been created you can get an Arc of the config by calling `get_info`.
+/// You can then start the run method, the fetcher will update the AgentInfoArc based on the
+/// given refresh interval.
+///
+/// # Example
+/// ```no_run
```

Comment:
I think the `no_run` doc tests are still compiled just not run (which is not possible since it does http calls). But I can add some test with the test agent.

Reply:
I added some integration tests with the test agent. We need to wait for https://github.com/DataDog/dd-apm-test-agent/pull/193 to get merged since the test agent does not send the agent state header. I also added a `require_test_agent` module to the trace-utils integration test to filter out all tests that use it in the pipeline.

Reply:
Ah, TIL. I incorrectly assumed `no_run` was completely ignored by the compiler. In that case, I have no objection to using it.

Reply:
> I also added a require_test_agent module to the trace-utils integration test to filter out all tests that use it in the pipeline.

I'm not sure that's necessary. We already run integration tests separately on CI. I think the correct separation for tests is the type (unit, integration, fuzz, etc). Also, by definition I believe all trace-utils integrations should be hitting the test agent anyway.

Reply:
We need a way to filter out the tests using the test agent in `.github/workflows/test.yml` since it requires docker, right ?

Reply:
Yes. We currently [do that](https://github.com/DataDog/libdatadog/blob/main/.github/workflows/test.yml#L57) with the `-E '!test(tracing_integration_tests::)'` flag.

Reply:
Can we assume all integration tests will require the test agent ?

Reply:
I think we can assume all integration tests may (but not necessarily) require external entities like containers running the test-agent or other services. I think we also want to assume integration tests should run separately from unit tests and other workflows in CI. These assumptions are why I think a less specific filter of just `integration_tests` is preferable to filtering on whether or not they use the test-agent. 

For data-pipeline or trace-utils specifically - yes I think we would assume the test-agent is being used. But I wouldn't necessarily say the same for other crates in libdatadog that may wish to write integration tests.

Reply:
I'm okay with running the integrations tests in a separated ci step. But we still need to be able to run the tests using the test agent separately as they only Linux (at least we [only run them on linux in the CI](https://github.com/DataDog/libdatadog/blob/09d1ef506d59c65cbe58f0a2e153d493716efaa3/.github/workflows/test.yml#L64)) whereas the other integration tests run on all platform.

Reply:
If you are worried about operating systems wouldn't it be better to just ignore tests on OSes we don't want to use with `#[cfg(not(target_os = "x"))]`?